src/java.desktop/windows/native/libawt/windows/ShellFolder2.cpp

```diff
@@ -1057,44 +1057,57 @@ JNIEXPORT jintArray JNICALL Java_sun_awt_shell_Win32ShellFolder2_getIconBits
             // Extract the color bitmap
             int nBits = iconSize * iconSize;
 
-            TRY;
-
-            long *colorBits = (long*)safe_Malloc(MAX_ICON_SIZE * MAX_ICON_SIZE * sizeof(long));
-            GetDIBits(dc, iconInfo.hbmColor, 0, iconSize, colorBits, &bmi, DIB_RGB_COLORS);
-            // XP supports alpha in some icons, and depending on device.
-            // This should take precedence over the icon mask bits.
-            BOOL hasAlpha = FALSE;
-            if (IS_WINXP) {
-                for (int i = 0; i < nBits; i++) {
-                    if ((colorBits[i] & 0xff000000) != 0) {
-                        hasAlpha = TRUE;
-                        break;
+            long *colorBits = NULL;
+            try {
+                entry_point();
+                colorBits = (long*)safe_Malloc(MAX_ICON_SIZE * MAX_ICON_SIZE * sizeof(long));
+            } catch(std::bad_alloc&) {
+                handle_bad_alloc();
+            }
+
+            if (colorBits != NULL) {
```

Comment:
I think this way defeats the use of exception to handle allocation error. You could use:
```c++
long *colorBits = (long*)malloc(MAX_ICON_SIZE * MAX_ICON_SIZE * sizeof(long));
if (colorBits != NULL) {
```
to achieve the same effect, which is shorter and clearer.

Reply:
Yes, but then handle_bad_alloc() will not be called which will not raise the OutOfMemoryError on the java side.

Reply:
> Yes, but then handle_bad_alloc() will not be called which will not raise the OutOfMemoryError on the java side.

Can't we call it explicitly in the case where `colorBits == NULL`?

Alternatively, since you don't use the macros for try and catch, you can re-arrange code and perform cleanup after the catch-block. Just an idea… It may not work.

Reply:
> Can't we call it explicitly in the case where colorBits == NULL?

I would use the default handlers for bad_alloc since they can be changed in feature and keeping code that is intended to be used in two places is usually not a good idea.

> Alternatively, since you don't use the macros for try and catch, you can re-arrange code and perform cleanup after the catch-block. Just an idea… It may not work.

That might require a lot of code reformatting and the benefits are unclear.

Reply:
> > Alternatively, since you don't use the macros for try and catch, you can re-arrange code and perform cleanup after the catch-block. Just an idea… It may not work.
> 
> That might require a lot of code reformatting and the benefits are unclear.

I tried it, it doesn't, [see this diff](https://github.com/aivanov-jdk/jdk/commit/ccf7500cf0e4614b12c0882204a81d928fce0a82). It's close to your latest code, however, there's only one `catch` that handles both allocations and there are no `if`'s inside the `try`-block.

Reply:
What about this approach?
[Compared to master](https://github.com/openjdk/jdk/compare/master...aivanov-jdk:8277299-bad-alloc-w32)
[Compared to your latest code](https://github.com/azuev-java/jdk/compare/JDK-8277299...aivanov-jdk:8277299-bad-alloc-w32)

I think it's cleaner in the use of exceptions. Each try-block allocates memory that it uses and frees it before the catch-block. There's no need for if's: if exception is thrown, the memory wasn't allocated, therefore the memory is freed only if no exception is thrown.

What do you think?

Reply:
I like the singular try/catch block approach better so i did that in the latest commit.